### DISCUSSION TITLE
pki tool generating RSA error

### DISCUSSION THREAD

**Discussion by JumboJa** [strongswan/strongswan#1984] [Q&A] [answered]
```
# pki --gen
no files found matching '/var/ipsec/strongswan.conf'
instantiation of DRBG_HMAC_SHA512 failed
building CRED_PRIVATE_KEY - RSA failed, tried 5 builders
private key generation failed

# pki --gen --type ed448
no files found matching '/var/ipsec/strongswan.conf'
building CRED_PRIVATE_KEY - ED448 failed, tried 2 builders
private key generation failed

# pki --gen --type ecdsa
no files found matching '/var/ipsec/strongswan.conf'
building CRED_PRIVATE_KEY - ECDSA failed, tried 2 builders
private key generation failed
```

ed25519 works OK though...
What could be a reason??
TNX

**Comment by tobiasbrunner** [comment]
Check what plugins are loaded (`pki --help`), look at the first error message for RSA, then read [this FAQ entry](https://docs.strongswan.org/docs/5.9/support/faq.html#_plugin_is_missing). Also consider that for ECDSA and EdDSA a plugin that provides these algorithms is necessary (e.g. _openssl_). For ed25519 you probably have the _ed25519_ plugin loaded.

**Comment by JumboJa** [comment]
TNX

```
pki --help
no files found matching '/var/ipsec/strongswan.conf'
strongSwan 5.9.5 PKI tool
loaded plugins: aes sha2 sha1 random x509 pubkey pkcs1 pkcs8 pem gmp curve25519 hmac
usage:
  pki --acert   (-z)  issue an attribute certificate
  pki --dn      (-d)  extract the subject DN of an X.509 certificate
  pki --gen     (-g)  generate a new private key
  pki --issue   (-i)  issue a certificate using a CA certificate and key
  pki --keyid   (-k)  calculate key identifiers of a key/certificate
  pki --pkcs12  (-u)  PKCS#12 functions
  pki --pkcs7   (-7)  PKCS#7 wrap/unwrap functions
  pki --print   (-a)  print a credential in a human readable form
  pki --pub     (-p)  extract the public key from a private key/certificate
  pki --req     (-r)  create a PKCS#10 certificate request
  pki --self    (-s)  create a self signed certificate
  pki --signcrl (-c)  issue a CRL using a CA certificate and key
  pki --verify  (-v)  verify a certificate using the CA certificate
  pki --help    (-h)  show usage information
```
I tried to get info from help previously, but I need to know what's the plugin name I need for RSA.

And I checked charon directory. All modules are set to load:

```
aes.conf
curve25519.conf
gcm.conf
gmp.conf
hmac.conf
kernel-netlink.conf
nonce.conf
pem.conf
pkcs1.conf
pkcs8.conf
pubkey.conf
random.conf
sha1.conf
sha2.conf
socket-default.conf
stroke.conf
updown.conf
vici.conf
x509.conf
xcbc.conf
```

As for openssl - I have openssl-util and libopenssl installed (OpenWRT), openssl generates RSA well. Also, 
openssl genpkey -algorithm ed25519 - OK
openssl ecparam -name secp256k1 -genkey - OK

However, pki doesn't ignite...

**Reply by tobiasbrunner** [reply]
The _gmp_ plugin, which handles RSA in this case (i.e. in the absence of another plugin that provides RSA) requires the _drbg_ plugin, which is enabled by default since 5.8.2. So why is it disabled in your case (see the FAQ entry)? Did you build strongSwan yourself? If so, what configure options did you use (and why)?

**Comment by JumboJa** [comment]
No, I used OpenWrt repo packages, not compiled myself. Also I didn't disable anything as well. And although gmp is loaded, there's no separate package containing "drbg". Seems there should be an option to enable it, as most likely it was compiled with gmp.
Or also another package is needed, seems gmp only is not enough.

```
swanctl --stats
loaded plugins: charon aes sha2 sha1 random nonce x509 pubkey pkcs1 pkcs8 pem gmp curve25519 xcbc hmac gcm kernel-netlink socket-default stroke vici updown
```

**Reply by tobiasbrunner** [reply, marked as answer]
As I said, the _gmp_ plugin requires the _drbg_ plugin. There might be a `strongswan-mod-drbg` package. Or you could perhaps install the `strongswan-mod-openssl` package as that pretty much can replace all the other crypto plugins.

**Reply by JumboJa** [reply]
Installed single package strongswan-mod-openssl, now it works. (strongswan-mod-drbg doesn't exist in Owrt repo, probably will compile it myself later)
TNX!!!
